src/backend/distributed/planner/distributed_planner.c

```diff
@@ -527,14 +543,44 @@ IsModifyDistributedPlan(DistributedPlan *distributedPlan)
 
 
 /*
- * CreateDistributedPlannedStmt encapsulates the logic needed to transform a particular
- * query into a distributed plan that is encapsulated by a PlannedStmt.
+ * CreateDistributedPlannedStmt encapsulates the logic needed to transform
+ * a particular query into a distributed plan that is encapsulated by a
+ * PlannedStmt.
  */
 static PlannedStmt *
 CreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan, Query *originalQuery,
 							 Query *query, ParamListInfo boundParams,
 							 PlannerRestrictionContext *plannerRestrictionContext)
 {
+	PlannedStmt *result = NULL;
+
+	if (QueryTreeContainsInlinableCTE(originalQuery))
+	{
+		/*
+		 * Inlining CTEs as subqueries in the query can avoid recursively
+		 * planning some (or all) of the CTEs. In other words, the inlined
+		 * CTEs could become part of query pushdown planning, which is much
+		 * more efficient than recursively planning. So, first try distributed
+		 * planning on the inlined CTEs in the query tree.
+		 *
+		 * We also should fallback to distributed planning with non-inlined CTEs
+		 * if the distributed planning fails with inlined CTEs, because recursively
+		 * planning CTEs can provide full SQL coverage, although it might be slow.
+		 */
+		result = InlineCtesAndCreateDistributedPlannedStmt(planId, localPlan,
```

Comment:
This looks kind of ok. I don't know whether the feature is worth the extra complexity, but it's something we could live with for a while.

Reply:
For some users, this change will help a lot, so as far as I can see, it is worth the complexity. I've seen quite a lot of performance gains for some cases, I'll summarize the results as well.